### ISSUE TITLE
Mutation and a query in the same time

### ISSUE THREAD

**Issue by drmax24** [Folkloreatelier/laravel-graphql#180]
It seems to me that i can not write a mutation and then a query in the same batch. Is it true?

```
mutation {
  userLogin(sessionToken: "9bcbb94d-ed35-4d0f-bcb6-e922cf587c15") {
    userId
  }
  someQuery {
    someId
  }
}
```

It doesn't work like this and it doesn't work no meter how i try

```
mutation {
  userLogin(sessionToken: "9bcbb94d-ed35-4d0f-bcb6-e922cf587c15") {
    userId
  }
}

query {
  someQuery {
    someId
  }
}
```

**Comment by Harry-Torry** [comment]
A query should use a `GET` request and a mutation should use a `POST` when using HTTP. This doesn't really apply when websockets are being used but I feel like trying to do a query and a mutation at the same time isn't the correct behaviour to be achieving.

**Comment by drmax24** [comment]
Let's say i write a computer game. I execute a batch of commands.
1. deploy nuke mutation
2. get environment cities
3. get environment rivers

Why it should not be possible?

**Comment by n1ru4l** [comment]
This package supports batches operation, you can send an array of
queries/mutation instead of a single mutation/query. Every single operation
will then be processed on their own.

**Comment by n1ru4l** [comment]
@Harry-Torry afaik there is no specific about how graphql should be handled over HTTP. GET Requests are can be cached way more easy. At the same time there is a limit for the length auf a URL, so this could become a problem with complex queries (but could be solved by persisted queries).

**Comment by Harry-Torry** [comment]
Oh cool, thanks guys! I was just wrong about everything!

**Comment by drmax24** [comment]
Very educating

**Comment by drmax24** [comment]
Well, i'm still thinking.
There was an example in another ticket
https://github.com/Folkloreatelier/laravel-graphql/issues/53

```
mutation something {
  loginWithToken(token: "6e37a03e-9ee4-42fd-912d-3f67d2d0d852"),
  do_stuff(greeting: "Hello", name: "Tom"),
  do_more_stuff(submarine_color: "Yellow")
}
```

Let's say do_stuff and do_more_stuff are not mutations but queries. Can you give a query example for this case?

**Comment by SonarSoftware** [comment]
query myQuery {
        mutation myMutation {
            //whatever
        }
        query someQuery {
            //whatever
        }
        query someOtherQuery {
            //whatever
        }
    }

**Comment by SonarSoftware** [comment]
You would have to use batching, I think https://dev-blog.apollodata.com/query-batching-in-apollo-63acfd859862 is a decent example of how it works in Apollo.

**Comment by drmax24** [comment]
We may mix terminology. Let's say execution of multiple instructions in single request to server.
Can i do it using graphiql?

```
mutation something {
  loginWithToken(token: "6e37a03e-9ee4-42fd-912d-3f67d2d0d852"),
  do_stuff(greeting: "Hello", name: "Tom"),
  do_more_stuff(submarine_color: "Yellow")
}
```
We will perform 3 operations here and will get 3 results. Is that an execution of multiple instructions? Yes.

But it seems that i can only do it without mixing mutations with queries

**Comment by SonarSoftware** [comment]
If they are three separate, unrelated functions, then they have to be batched.

If all three of those things need to happen as part of the 'loginWithToken' migration, you would just implement that functionality into the 'loginWithToken' migration. 

There is no way to run three, separate, unrelated queries/migrations without using batching.

**Comment by drmax24** [comment]
Yes 3 unrelated operations. But in apollo post it explains the concept. I don't really understand how to write a request in my graphiql query builder

**Comment by SonarSoftware** [comment]
Yeah, good question - I am not sure you are actually able to batch both queries and mutations in a single request, maybe they do need to be split. I don't implement the batching by hand in my application, I just use Apollo and let it handle it. If true, then this would be a limitation of GraphQL rather than the library itself.

**Comment by drmax24** [comment]
This is the question of my entire post. This should be possible but it is not somewhy.

**Comment by SonarSoftware** [comment]
Where are you seeing in the spec that it's supported?

**Comment by drmax24** [comment]
Well you just gave me a link many times explaining batching concept. However in my very first post i had the same batching query. Now you ask why do we need batching at all?

**Comment by SonarSoftware** [comment]
No, all I'm saying is I misunderstood what you were trying to do. It doesn't appear batching allows mixing of queries and mutations. My response was to you saying 'it should be possible' - from what I am reading in the GraphQL batching spec, it does not appear that it is possible to mix both mutations and queries.

Batching allows you to make multiple requests in a single HTTP request, yes. It does not appear to allow you to mix both queries and mutations in a single request. The only solution I would see would be to expose the data you want to return in the query as part of the return value of the mutation, in which case you could encapsulate your secondary queries as the mutation return.

**Comment by drmax24** [comment]
I want to achieve this:
https://github.com/Folkloreatelier/laravel-graphql/issues/180#issuecomment-323706640

**Comment by SonarSoftware** [comment]
I understand. I don't think that is possible per the GraphQL spec. The only solution I would see would be to have the mutation itself return a type, which is a list of the users, cities, and rivers types, and then you would refer to those in the query return section of the mutation. I don't think there is any way to just run an arbitrary query in the same request as the mutation.

**Comment by drmax24** [comment]
I think this is a mistake. Sounds like a self-castration. I will write all mutations as queries in this case.

**Comment by SonarSoftware** [comment]
Just bear in mind the spec does not dictate the execution order of multiple queries, so if there is some requirement of prior execution for a query based on the position in the request, you may run into future failures by following that approach. This is the big difference between mutations and queries - mutations are resolved in series, but queries are resolved in parallel.

**Comment by drmax24** [comment]
Thank you for info. That's interesting. Php implementation does not have any multithreading i'm sure

**Comment by drmax24** [comment]
Still hoping that mixing of queries and mutations is or would be possible

**Comment by SonarSoftware** [comment]
Might be in the future, but it would have to be something supported in the GraphQL spec, it wouldn't be specific to this library.

**Comment by n1ru4l** [comment]
@SonarSoftware I do not think batching is related to the spec at all, so this is no limitation. Also the implementation in laravel-graphql does not check for mixed batched results. 

**Therefore batching is totally okay with mixed queries and mutations.** @drmax24

**Comment by drmax24** [comment]
Ah thank you @n1ru4l.
Our company writes business apps using this lib. And i've started a computer game in a spare time. Imagine effectiveness for of this approach for computer games. 

But @n1ru4l. I've tried mixing queries and mutations i get errors. Can you look into that? Just try to run one query and one mutation in the same time

**Comment by n1ru4l** [comment]
@drmax24 You can not mix queries and mutations in a single operation. Since laravel-graphql support query batching you can send an array of operations:

```text
[
  {
    query: "< query 0 >",
    variables:  < variables for query 0 >
  },
  {
    query: "< query 1 >",
    variables: < variables for query 1 >
  },
  ...
  {
    query: "< query n >",
    variables: < variables for query n >
  }
]
```

And will get something like this in return:

```text
[
  <result for query 0>,
  <result for query 1>,
  ...
  <result for query n>
]
```

**Comment by drmax24** [comment]
Can i do that in graphiql?

**Comment by n1ru4l** [comment]
@drmax24 nope you can not

**Comment by drmax24** [comment]
I think that may make integration harder for front-end developer. Because if you can not write a batch in a get request than it's getting harder.
You've mentioned apollo. Is that another client like graphql?

**Comment by n1ru4l** [comment]
@drmax24 Apollo client is an extensible GraphQL client like Relay with support for all major frontend frameworks like React, Angular and Vue.js.

GraphiQL is a Devtoolish gui for getting to know your graphql endpoint.

You could either test every single operation with graphiQL, or if thats an issue for you convert all your queries to mutations (which I personally would not do).

**Comment by drmax24** [comment]
Graphiql = GET format
Can i write "mixed" batch request as a GET request? Non-mixed i can

**Comment by n1ru4l** [comment]
@drmax24 The graphiql packaged with laravel-graphql uses POST requests.

**Comment by drmax24** [comment]
Graphiql gives get varian supereasy. Well. Need to figure out if i can use get and if i can not why

**Comment by n1ru4l** [comment]
If any of your operations includes a mutation you should use POST anyway because you do not want to expose possible sensitive information via GET params. I am pretty sure it is not possible to send batchOperations with laravel-graphql and any other graphql server implementation.

**Comment by drmax24** [comment]
Get allows you to share query between developers much easier. We stick to get to minimize development time.

**Comment by n1ru4l** [comment]
You will have to override the GraphQL Controller to make this possible

**Comment by SonarSoftware** [comment]
> I do not think batching is related to the spec at all, so this is no limitation. Also the implementation in laravel-graphql does not check for mixed batched results.

It is not related to the spec as it exists at graphql.org, but it was designed and implemented by Facebook and described at React Europe. That being said, there are a number of pieces of it that haven't been implemented in any libraries to my knowledge, like the ability to define relationships between the batched queries.

I'm not saying it wouldn't be a useful feature, I can see that it would (and I think you can overcome the obstacle of parallel vs series by just defining that mutations always run first) - my answer was to say that it isn't a limitation of the library, it just isn't defined behavior in GraphQL.

That all being said, I am not the author of the library, but so far, it seems to have followed closely the webonyx/graphql library, simply wrapping it into Laravel, and that implementation is a port of the reference GraphQL spec to PHP. 

n1ru4l is 100% correct though - if you really want to implement this yourself, it is probably not that hard to do, you would just have to decide the format of mixed queries and mutations, and follow through the existing code. I think all the execution from my recollection actually happens in the webonyx parts, rather than this library, but it should be doable.

**Comment by n1ru4l** [comment]
@SonarSoftware I just watched parts of the video and read the recap on the apollo blog and now I understand how these things are meant.

The batching is handled in the GraphQL controller of laravel-graphql, since you can define your own Controller for handling the operations this can be easily achieved (as you said).

@drmax24 Sharing an encoded uri string does not seem that convenient to me 🤔. However I think you should be capable of overwriting the Controller to do this. I think nobody would be against a PR for this feature.

**Comment by drmax24** [comment]
Here is an example that can be run in graphiql
https://github.com/rebing/graphql-laravel/blob/master/docs/advanced.md#batching

**Comment by SonarSoftware** [comment]
Nice, does this work in this library or just in the rebing one? 

Would be interesting to see what the implementation is here to ensure query consistency (e.g. are the mutations run first, or is it all in parallel.)

**Comment by n1ru4l** [comment]
@SonarSoftware Relevant Code: https://github.com/Folkloreatelier/laravel-graphql/blob/master/src/Folklore/GraphQL/GraphQLController.php#L10

ATM they are processed one after the other

**Comment by drmax24** [comment]
Folkoreatelier/laravel-graphql can't run above query.
Rebing has code for controller:
https://github.com/rebing/graphql-laravel/blob/master/src/Rebing/GraphQL/GraphQLController.php

About parallel execution:
I think queries should run in parallel with each other but in sequence with mutations.
However we will never have parallel execution on php libraries because of language limitations.

**Comment by dmongeau** [comment]
Could you test this with 1.0.23. Should work

**Comment by Vokinloksar** [comment]
This is a very important thread on batching queries and mutations I think, but the _batching_ method mentioned [above](https://github.com/Folkloreatelier/laravel-graphql/issues/180#issuecomment-324127765)  is async, what if I want to send a mutation and a query afterwards in sequence? Is there any way?
